File: android/app/src/main/java/com/brush/BrushGeofenceModule.kt
/**
 * BrushGeofenceModule.kt — Native geofence wrapper for Android (KAN-56).
 *
 * Wraps GeofencingClient (Google Play Services) so React Native JS can
 * register and remove circular geofences without a third-party library.
 *
 * Supports:
 *   - registerGeofence(id, lat, lng, radiusMeters)
 *   - removeGeofence(id)
 *   - removeAllGeofences()
 *   - Emits 'onGeofenceEntry' to JS via GeofenceBroadcastReceiver
 *
 * Android limits: 100 geofences per app. With ≤10 POI types in practice
 * this limit is not a concern.
 *
 * Note: ACCESS_BACKGROUND_LOCATION is required on Android 10+ for background
 * geofence callbacks. This permission is already requested in geolocation.ts
 * (KAN-22).
 */

package com.brush

import android.app.PendingIntent
import android.content.Intent
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.google.android.gms.location.*

class BrushGeofenceModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val geofencingClient: GeofencingClient =
        LocationServices.getGeofencingClient(reactContext)

    private val geofencePendingIntent: PendingIntent by lazy {
        val intent = Intent(reactContext, GeofenceBroadcastReceiver::class.java)
        PendingIntent.getBroadcast(
            reactContext,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE,
        )
    }

    override fun getName() = "BrushGeofenceModule"

    // ── JS-callable methods ───────────────────────────────────────────────────

    @ReactMethod
    fun registerGeofence(
        id: String,
        lat: Double,
        lng: Double,
        radiusMeters: Double,
        promise: Promise,
    ) {
        val geofence = Geofence.Builder()
            .setRequestId(id)
            .setCircularRegion(lat, lng, radiusMeters.toFloat())
            .setExpirationDuration(Geofence.NEVER_EXPIRE)
            .setTransitionTypes(Geofence.GEOFENCE_TRANSITION_ENTER)
            .build()

        val request = GeofencingRequest.Builder()
            .setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER)
            .addGeofence(geofence)
            .build()

        geofencingClient.addGeofences(request, geofencePendingIntent)
            .addOnSuccessListener { promise.resolve(null) }
            .addOnFailureListener { e -> promise.reject("GEOFENCE_ERROR", e.message, e) }
    }

    @ReactMethod
    fun removeGeofence(id: String, promise: Promise) {
        geofencingClient.removeGeofences(listOf(id))
            .addOnSuccessListener { promise.resolve(null) }
            .addOnFailureListener { e -> promise.reject("GEOFENCE_ERROR", e.message, e) }
    }

    @ReactMethod
    fun removeAllGeofences(promise: Promise) {
        geofencingClient.removeGeofences(geofencePendingIntent)
            .addOnSuccessListener { promise.resolve(null) }
            .addOnFailureListener { e -> promise.reject("GEOFENCE_ERROR", e.message, e) }
    }

    // ── Event emitter ─────────────────────────────────────────────────────────

    /**
     * Called by GeofenceBroadcastReceiver when a GEOFENCE_TRANSITION_ENTER
     * event is received. Forwards the geofence ID to JS.
     */
    fun sendGeofenceEntryEvent(geofenceId: String) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onGeofenceEntry", Arguments.createMap().apply {
                putString("geofenceId", geofenceId)
            })
    }

    // Required by RCTEventEmitter interface
    @ReactMethod fun addListener(eventName: String) {}
    @ReactMethod fun removeListeners(count: Int) {}
}

File: ios/Brush/BrushGeofenceModule.swift
/**
 * BrushGeofenceModule.swift — Native geofence wrapper for iOS (KAN-56).
 *
 * Wraps CLLocationManager's region monitoring API so React Native JS can
 * register and remove circular geofences without a third-party library.
 *
 * Supports:
 *   - registerGeofence(id, lat, lng, radiusMeters)
 *   - removeGeofence(id)
 *   - removeAllGeofences()
 *   - Emits 'onGeofenceEntry' event to JS on CLCircularRegion entry
 *
 * iOS limits: 20 monitored regions per app. With ≤10 POI types in practice
 * this limit is not a concern.
 *
 * Note: CLCircularRegion boundary crossing simulation does NOT work in the
 * iOS Simulator. Test on a real device or use Instruments → Location simulation.
 */

import Foundation
import CoreLocation
import React

@objc(BrushGeofenceModule)
class BrushGeofenceModule: RCTEventEmitter, CLLocationManagerDelegate {

  private let locationManager = CLLocationManager()
  private var hasListeners = false

  // MARK: - RCTEventEmitter

  override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String]! {
    return ["onGeofenceEntry"]
  }

  override func startObserving() {
    hasListeners = true
  }

  override func stopObserving() {
    hasListeners = false
  }

  // MARK: - Lifecycle

  override init() {
    super.init()
    locationManager.delegate = self
    // 'always' authorisation is required for background geofence callbacks.
    // The permission was already requested in geolocation.ts (KAN-22).
    locationManager.requestAlwaysAuthorization()
  }

  // MARK: - JS-callable methods

  @objc func registerGeofence(
    _ id: String,
    lat: Double,
    lng: Double,
    radiusMeters: Double,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let center     = CLLocationCoordinate2D(latitude: lat, longitude: lng)
    let radius     = min(radiusMeters, locationManager.maximumRegionMonitoringDistance)
    let region     = CLCircularRegion(center: center, radius: radius, identifier: id)
    region.notifyOnEntry = true
    region.notifyOnExit  = false // exit events not needed for v1 notification flow

    locationManager.startMonitoring(for: region)
    resolve(nil)
  }

  @objc func removeGeofence(
    _ id: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    for region in locationManager.monitoredRegions where region.identifier == id {
      locationManager.stopMonitoring(for: region)
    }
    resolve(nil)
  }

  @objc func removeAllGeofences(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    for region in locationManager.monitoredRegions {
      locationManager.stopMonitoring(for: region)
    }
    resolve(nil)
  }

  // MARK: - CLLocationManagerDelegate

  func locationManager(_ manager: CLLocationManager, didEnterRegion region: CLRegion) {
    guard hasListeners, let circularRegion = region as? CLCircularRegion else { return }
    sendEvent(withName: "onGeofenceEntry", body: ["geofenceId": circularRegion.identifier])
  }

  func locationManager(_ manager: CLLocationManager, monitoringDidFailFor region: CLRegion?, withError error: Error) {
    print("[BrushGeofenceModule] Monitoring failed for region \(region?.identifier ?? "unknown"): \(error.localizedDescription)")
  }
}
